packages/mobile/src/screens/chat-screen/ChatScreen.tsx

```diff
@@ -308,9 +339,80 @@ export const ChatScreen = () => {
     )
   }
 
+  const closeReactionPopup = useCallback(() => {
+    setShouldShowPopup(false)
+    setPopupChatIndex(null)
+  }, [setShouldShowPopup, setPopupChatIndex])
+
+  const handleReactionSelected = useCallback(
+    (message: Nullable<ChatMessage>, reaction: ReactionTypes) => {
+      if (userId && message) {
+        dispatch(
+          setMessageReaction({
+            userId,
+            chatId,
+            messageId: message.message_id,
+            reaction:
+              message.reactions?.find((r) => r.user_id === userIdEncoded)
+                ?.reaction === reaction
+                ? null
+                : reaction
+          })
+        )
+      }
+      closeReactionPopup()
+    },
+    [dispatch, userIdEncoded, chatId, userId, closeReactionPopup]
+  )
+
+  // Calculates positions of reactions popup and highlighted chat message to display
+  // in the portal.
+  const handleMessagePress = async (index: number) => {
+    if (index < 0 || index >= chatMessages.length) {
+      return
+    }
+    const popupChatMessage = chatMessages[index]
+    const popupViewRef = itemsRef.current[index]
+    if (popupViewRef === null || popupViewRef === undefined) {
+      return
+    }
+    setPopupChatIndex(index)
+    setPopupChatHasTail(hasTail(popupChatMessage, chatMessages[index - 1]))
+    const senderUserId = decodeHashId(popupChatMessage.sender_user_id)
+    setPopupIsAuthor(senderUserId === userId)
+    const selectedReactionValue = popupChatMessage.reactions?.find(
+      (r) => r.user_id === encodeHashId(userId)
+    )?.reaction
+    if (selectedReactionValue) {
+      setSelectedReaction(selectedReactionValue)
+    }
+
+    // Measure position of selected message to create a copy of it on top
+    // of the dimmed background inside the portal.
+    const { top: messageY, height: messageHeight } = await new Promise<{
+      top: number
+      height: number
+    }>((resolve) => {
+      popupViewRef.measureInWindow((x, y, width, height) => {
+        resolve({ top: y, height })
+      })
+    })
+
+    // Determine whether to show reactions popup above or below highlighted message.
+    const spaceAboveMessage = messageY - rootY
+    const shouldShowReactionsAbove =
+      spaceAboveMessage > REACTION_CONTAINER_HEIGHT
+    const reactionY = shouldShowReactionsAbove
+      ? messageY - REACTION_CONTAINER_HEIGHT - rootY + spacing(3.5)
+      : messageY + messageHeight - rootY + spacing(2)
+    setReactionY(reactionY)
+    setMessageY(messageY - rootY)
+    setShouldShowPopup(true)
```

Comment:
Since many of these values change together, I would support grouping them in an object-based useStates as it makes sense. React docs do advocate for grouping values that change together in useState: https://reactjs.org/docs/hooks-faq.html#should-i-use-one-or-many-state-variables

cc @rickyrombo

Reply:
Some do change together like messageY and reactionY but rootY and composeY are independent from the rest and each other.. I'm liking the ref idea instead bc none of these should trigger a re-render.